File: android/src/main/java/expo/modules/smartrefreshlayout/ExpoSmartrefreshlayoutModule.kt
package expo.modules.smartrefreshlayout

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoSmartrefreshlayoutModule : Module() {
    override fun definition() = ModuleDefinition {

        Name("ExpoSmartrefreshlayout")


        // 注册 Expo View
        View(ExpoSmartrefreshlayoutView::class) {

            // ------------------- Props -------------------
            // Header/Footer 高度、拖拽、触发比率
            Prop<Double>("dragRate") { view, value -> view.setDragRate(value) }
            Prop<Double>("headerHeight") { view, value -> view.setHeaderHeight(value) }
            Prop<Double>("footerHeight") { view, value -> view.setFooterHeight(value) }
            Prop<Double>("headerInsetStart") { view, value -> view.setHeaderInsetStart(value) }
            Prop<Double>("footerInsetStart") { view, value -> view.setFooterInsetStart(value) }
            Prop<Double>("headerMaxDragRate") { view, value -> view.setHeaderMaxDragRate(value) }
            Prop<Double>("footerMaxDragRate") { view, value -> view.setFooterMaxDragRate(value) }
            Prop<Double>("headerTriggerRate") { view, value -> view.setHeaderTriggerRate(value) }
            Prop<Double>("footerTriggerRate") { view, value -> view.setFooterTriggerRate(value) }
            Prop<Int>("reboundDuration") { view, value -> view.setReboundDuration(value) }

            // Header 类型
            Prop<String>("headerType") { view, value -> view.setHeaderType(value) }

            // 启用/禁用各种功能
            Prop<Boolean>("enableRefresh") { view, value -> view.setEnableRefresh(value) }
            Prop<Boolean>("enableLoadMore") { view, value -> view.setEnableLoadMore(value) }
            Prop<Boolean>("enableAutoLoadMore") { view, value -> view.setEnableAutoLoadMore(value) }
            Prop<Boolean>("enablePureScrollMode") { view, value -> view.setEnablePureScrollMode(value) }
            Prop<Boolean>("enableScrollContentWhenLoaded") { view, value -> view.setEnableScrollContentWhenLoaded(value) }
            Prop<Boolean>("enableScrollContentWhenRefreshed") { view, value -> view.setEnableScrollContentWhenRefreshed(value) }
            Prop<Boolean>("enableOverScrollDrag") { view, value -> view.setEnableOverScrollDrag(value) }
            Prop<Boolean>("enableOverScrollBounce") { view, value -> view.setEnableOverScrollBounce(value) }
            Prop<Boolean>("enableNestedScroll") { view, value -> view.setEnableNestedScroll(value) }
            Prop<Boolean>("enableHeaderTranslationContent") { view, value -> view.setEnableHeaderTranslationContent(value) }
            Prop<Boolean>("enableFooterTranslationContent") { view, value -> view.setEnableFooterTranslationContent(value) }
            Prop<Boolean>("enableLoadMoreWhenContentNotFull") { view, value -> view.setEnableLoadMoreWhenContentNotFull(value) }

            // Classic Header/Footer props
            Prop("classicRefreshHeaderProps") { view: ExpoSmartrefreshlayoutView, value: Map<String, Any?> -> 
                view.setClassicRefreshHeaderProps(value)
            }
            Prop("classicLoadMoreFooterProps") { view: ExpoSmartrefreshlayoutView, value: Map<String, Any?> -> 
                view.setClassicLoadMoreFooterProps(value)
            }
            
            // Custom Header/Footer flags
            Prop<Boolean>("hasCustomHeader") { view, value -> view.setHasCustomHeader(value) }
            Prop<Boolean>("hasCustomFooter") { view, value -> view.setHasCustomFooter(value) }
            
            // Haptic Feedback
            Prop<Boolean>("enableHapticFeedback") { view, value -> view.setEnableHapticFeedback(value) }


            // ------------------- Events -------------------
            Events(
                "onRefresh",
                "onLoadMore",
                "onStateChanged",
                "onHeaderMoving",
                "onFooterMoving"
            )

            // ------------------- Commands -------------------

            AsyncFunction("finishRefresh") { view: ExpoSmartrefreshlayoutView, success: Boolean, delay: Int ->
                view.finishRefresh(success, delay)
            }

            AsyncFunction("finishLoadMore") { view: ExpoSmartrefreshlayoutView, success: Boolean, delay: Int, noMoreData: Boolean ->
                view.finishLoadMore(success, delay, noMoreData)
            }

            AsyncFunction("autoRefresh") { view: ExpoSmartrefreshlayoutView, delay: Int ->
                view.autoRefresh(delay)
            }

            AsyncFunction("autoLoadMore") { view: ExpoSmartrefreshlayoutView, delay: Int ->
                view.autoLoadMore(delay)
            }

            AsyncFunction("setNoMoreData") { view: ExpoSmartrefreshlayoutView, noMore: Boolean ->
                view.setNoMoreData(noMore)
            }
        }
    }
}

File: ios/ExpoSmartrefreshlayoutModule.swift
import ExpoModulesCore

public class ExpoSmartrefreshlayoutModule: Module {
    public func definition() -> ModuleDefinition {
        // 模块名称
        Name("ExpoSmartrefreshlayout")
        
        // 定义事件
        Events(
            "onRefresh",
            "onLoadMore",
            "onStateChanged",
            "onHeaderMoving",
            "onFooterMoving"
        )
        
        // 注册 View 组件
        View(ExpoSmartrefreshlayoutView.self) {
            
            // ------------------- Events -------------------
            Events(
                "onRefresh",
                "onLoadMore",
                "onStateChanged",
                "onHeaderMoving",
                "onFooterMoving"
            )
            
            // ------------------- Props -------------------
            
            // 基础配置
            Prop("enableRefresh") { (view: ExpoSmartrefreshlayoutView, value: Bool) in
                view.setEnableRefresh(value)
            }
            
            Prop("enableLoadMore") { (view: ExpoSmartrefreshlayoutView, value: Bool) in
                view.setEnableLoadMore(value)
            }
            
            Prop("enableAutoLoadMore") { (view: ExpoSmartrefreshlayoutView, value: Bool) in
                view.setEnableAutoLoadMore(value)
            }
            
            // Header 类型
            Prop("headerType") { (view: ExpoSmartrefreshlayoutView, value: String) in
                view.setHeaderType(value)
            }
            
            // 经典样式配置
            Prop("classicRefreshHeaderProps") { (view: ExpoSmartrefreshlayoutView, value: [String: Any]) in
                view.setClassicRefreshHeaderProps(value)
            }
            
            Prop("classicLoadMoreFooterProps") { (view: ExpoSmartrefreshlayoutView, value: [String: Any]) in
                view.setClassicLoadMoreFooterProps(value)
            }
            
            // 自定义 Header 标志
            Prop("hasCustomHeader") { (view: ExpoSmartrefreshlayoutView, value: Bool) in
                view.setHasCustomHeader(value)
            }
            
            // 触觉反馈
            Prop("enableHapticFeedback") { (view: ExpoSmartrefreshlayoutView, value: Bool) in
                view.setEnableHapticFeedback(value)
            }
        }
        
        // ------------------- Commands/Functions -------------------
        
        /**
         * 完成刷新操作
         * @param view 视图实例
         * @param success 是否刷新成功
         * @param delay 延迟时间（毫秒）
         */
        Function("finishRefresh") { (view: ExpoSmartrefreshlayoutView, success: Bool, delay: Int) in
            view.finishRefresh(success: success, delay: delay)
        }
        
        /**
         * 完成加载更多操作
         * @param view 视图实例
         * @param success 是否加载成功
         * @param delay 延迟时间（毫秒）
         * @param noMoreData 是否没有更多数据
         */
        Function("finishLoadMore") { (view: ExpoSmartrefreshlayoutView, success: Bool, delay: Int, noMoreData: Bool) in
            view.finishLoadMore(success: success, delay: delay, noMoreData: noMoreData)
        }
        
        /**
         * 自动刷新
         * @param view 视图实例
         * @param delay 延迟时间（毫秒）
         */
        Function("autoRefresh") { (view: ExpoSmartrefreshlayoutView, delay: Int) in
            view.autoRefresh(delay: delay)
        }
        
        /**
         * 自动加载更多
         * @param view 视图实例
         * @param delay 延迟时间（毫秒）
         */
        Function("autoLoadMore") { (view: ExpoSmartrefreshlayoutView, delay: Int) in
            view.autoLoadMore(delay: delay)
        }
        
        /**
         * 设置是否没有更多数据
         * @param view 视图实例
         * @param noMore 是否没有更多数据
         */
        Function("setNoMoreData") { (view: ExpoSmartrefreshlayoutView, noMore: Bool) in
            view.setNoMoreData(noMore: noMore)
        }
    }
}
